### PR TITLE
feat: add web search and provider routing support

### PR DESCRIPTION
**Pull request by louisgv**
# feat: add web search and provider routing support

## Summary

This PR adds comprehensive support for OpenRouter's web search and provider routing features to the AI SDK provider. The implementation includes:

**Web Search Support:**
- Plugin-based web search via `plugins` array with `web` type
- Built-in web search via `web_search_options` parameter  
- Automatic web search enablement for models with `:online` suffix
- Web search annotation processing in both streaming and non-streaming responses

**Provider Routing Support:**
- Complete provider routing configuration including `order`, `allow_fallbacks`, `require_parameters`, `data_collection`
- Advanced filtering options: `only`, `ignore`, `quantizations`, `sort`, `max_price`
- Support for all documented quantization levels and sorting criteria

The implementation maintains backward compatibility and follows existing code patterns for parameter handling and response processing.

## Review & Testing Checklist for Human

- [ ] **Test web search functionality end-to-end** - Verify both plugin-based (`plugins: [{ type: 'web' }]`) and built-in (`web_search_options`) approaches work with real API calls
- [ ] **Verify provider routing parameters** - Test that provider routing options like `order`, `sort`, `quantizations`, etc. are properly sent to the API and affect routing behavior
- [ ] **Check `:online` model suffix handling** - Confirm that using a model ID ending in `:online` automatically enables web search when no explicit plugins are provided
- [ ] **Validate annotation processing** - Ensure web search annotations appear correctly in responses and don't break existing content handling
- [ ] **Test for regressions** - Verify existing chat completion functionality (without web search/routing) still works correctly

**Recommended test plan:**
1. Test basic web search: `openrouter('meta-llama/llama-3.1-8b-instruct:online', { ... })`
2. Test plugin configuration: `plugins: [{ type: 'web', max_results: 3, search_prompt: 'Find recent news' }]`
3. Test provider routing: `provider: { order: ['anthropic', 'openai'], sort: 'price' }`
4. Verify streaming responses include annotations correctly
5. Test edge cases like empty annotations, malformed URLs, etc.

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    User[User Code] --> Settings["src/types/openrouter-chat-settings.ts"]:::major-edit
    Settings --> ChatModel["src/chat/index.ts"]:::major-edit
    ChatModel --> Schemas["src/chat/schemas.ts"]:::major-edit
    ChatModel --> API[OpenRouter API]:::context
    API --> Responses[API Responses]:::context
    Responses --> ChatModel
    
    Settings --> WebSearch["Web Search Types<br/>plugins, web_search_options"]:::major-edit
    Settings --> Routing["Provider Routing Types<br/>order, sort, quantizations"]:::major-edit
    
    ChatModel --> GetArgs["getArgs() method<br/>:online suffix handling"]:::major-edit
    ChatModel --> Processing["Response Processing<br/>annotations handling"]:::major-edit
    
    Schemas --> Annotations["Annotation Schemas<br/>url_citation objects"]:::major-edit

    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit    
        L3[Context/No Edit]:::context
    end

    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- **Session Info**: Requested by Louis (@louisgv) - Session: https://app.devin.ai/sessions/1eae218929d5447b867ff4fd78634672
- **Type Safety**: All new types are optional to maintain backward compatibility
- **Documentation**: Types include comprehensive JSDoc comments explaining each parameter
- **Testing**: All existing tests pass, but new functionality requires manual API testing
- **Risk Level**: Medium-High - Substantial new functionality that couldn't be fully validated without live API access

The implementation replicates OpenRouter's documented API structure but should be validated against actual API behavior, especially for edge cases in web search annotation handling and provider routing parameter validation.

### CODE REVIEW

File: src/chat/index.ts [hunk 1]
```diff
@@ -93,6 +93,10 @@ export class OpenRouterChatLanguageModel implements LanguageModelV2 {
       model: this.modelId,
       models: this.settings.models,
 
+      ...(this.modelId.endsWith(':online') && !this.settings.plugins && {
+        plugins: [{ type: 'web' as const }]
+      }),
```
Comment (louisgv) [inline src/chat/index.ts:98]: remove this

File: src/chat/index.ts [hunk 2]
```diff
@@ -318,6 +329,17 @@ export class OpenRouterChatLanguageModel implements LanguageModelV2 {
...
+    if (choice.message.annotations) {
+      for (const annotation of choice.message.annotations) {
+        if (annotation.type === 'url_citation') {
+          content.push({
+            type: 'text' as const,
+            text: `[Citation: ${annotation.title || annotation.url}](${annotation.url})`,
+          });
+        }
+      }
+    }
```
Comment (louisgv) [inline src/chat/index.ts:341]: Check if the aisdk has better data type for citation/annotation

File: src/chat/schemas.ts [hunk 3]
```diff
@@ -103,6 +117,20 @@ export const OpenRouterStreamChatCompletionChunkSchema = z.union([
...
+             * Web search annotations containing citation information
+             */
+            annotations: z.array(
+              z.object({
+                type: z.literal('url_citation'),
+                url: z.string(),
+                title: z.string().optional(),
+                content: z.string().optional(),
+                start_index: z.number().optional(),
+                end_index: z.number().optional(),
```
Comment (louisgv) [inline src/chat/schemas.ts:131]: wrong schema -- this should be:

```
  url_citation: z.object({
    end_index: z.number(),
    start_index: z.number(),
    title: z.string(),
    url: z.string(),
    content: z.string().optional(),
  }),

```
